### PR TITLE
feat(agent): cross-chat context retrieval for same-user threads (#1505)

### LINKED ISSUE #1505: Agent does not use context from other chats by the same user
## Summary

The agent can use context from earlier messages in the same chat, but it does not retrieve relevant context from other chats belonging to the same user.

## Problem

When a user continues a topic in the same chat, the agent appears to understand prior messages. However, if the same user has relevant context in another chat, the agent does not pick it up and behaves as if that context does not exist.

Expected behavior: user-level memory and chat history retrieval should allow the agent to find relevant context across chats for the same user, while still respecting privacy, workspace, and source boundaries.

Impact: users have to repeat information across conversations, and the agent’s personalization/memory feels limited to the current thread instead of the user’s broader history.

Steps to reproduce:
1. Start Chat A as a user and share a durable fact, preference, or project context.
2. Start Chat B as the same user.
3. Ask a question or make a request in Chat B that depends on the context from Chat A.
4. Observe that the agent does not use the relevant context from Chat A.
5. Ask the same dependent question later inside Chat A.
6. Observe that the agent can use the previous same-chat context.

Version / platform: OpenHuman app / agent memory behavior; exact build and platform to confirm.

## Solution (optional)

Investigate the memory and conversation retrieval path to confirm whether retrieval is scoped only to the active chat/session. Add or fix user-level retrieval so relevant memories and prior chat summaries from the same user can be considered across chats, with source/session provenance and privacy boundaries preserved.

## Acceptance criteria

- [ ] **Repro gone** — The agent can retrieve relevant context from another chat belonging to the same user when it is useful for the current request.
- [ ] **Same-chat behavior preserved** — Existing same-chat context handling continues to work.
- [ ] **User scope enforced** — Cross-chat retrieval only uses chats/memories belonging to the same user and permitted workspace/source scope.
- [ ] **Relevance bounded** — The agent retrieves relevant cross-chat context without dumping unrelated conversation history into prompts.
- [ ] **Provenance available** — Retrieved cross-chat context keeps enough provenance to debug which chat/source supplied it.
- [ ] **Privacy safe** — Cross-chat context does not leak across users, accounts, workspaces, or disconnected sources.
- [ ] **Regression safety** — Unit, integration, or E2E coverage verifies same-user cross-chat retrieval, same-chat retrieval, and cross-user isolation.
- [ ] **Diff coverage ≥ 80%** — the fix PR meets the changed-lines coverage gate (Vitest + cargo-llvm-cov, enforced by [`.github/workflows/coverage.yml`](../../.github/workflows/coverage.yml)).

## Related

- Memory retrieval and user personalization workstream.

### PR DESCRIPTION
**Pull request by oxoxDev**
## Summary

- Adds `[Cross-chat context]` block to the agent prompt drawn from the same user's other JSONL chat threads, with provenance tags.
- New `ConversationStore::search_cross_thread_messages` scans the user's `workspace/memory/conversations/threads/*.jsonl` for relevance-scored substring matches, excluding the active thread.
- `DefaultMemoryLoader` gains an optional `workspace_dir`; when wired (via `with_workspace_dir`), it consults the JSONL scan directly, bypassing the FTS5 path that currently fails to populate for live chats.
- `RecallOpts.cross_session` + `UnifiedMemory::episodic_cross_session_search` retained as the fallback path for tests / headless callers without `workspace_dir`.
- Session builder wires `config.workspace_dir` into the loader so live chats hit the JSONL scan path.

## Problem

The agent retrieves prior context from the active chat but not from the same user's other chats. Users have to repeat themselves across threads.

Symptom: the cross-session path used by `DefaultMemoryLoader` reads `episodic_log` (FTS5), which is only populated by the explicit `archive_session` tool. Live conversation content lives in per-thread JSONL files and never reached the prompt.

## Solution

Two complementary paths:

1. **Primary (live chats):** JSONL substring scan over the user's `workspace/memory/conversations/threads/*.jsonl`, scoring each message by the fraction of query terms it contains (terms ≥ 3 chars). Hits are filtered by `min_relevance_score`, capped at `CROSS_CHAT_LIMIT`, snippet-truncated, and rendered as `- [<provenance>] <snippet>` lines under a `[Cross-chat context]` header in the assembled prompt. Active thread is excluded via `thread_context::current_thread_id()`.

2. **Fallback (tests / headless):** when `workspace_dir` is not wired, fall back to `Memory::recall(query, opts.cross_session=true)`, which calls into the existing `UnifiedMemory::episodic_cross_session_search` FTS5 helper.

Privacy / scope: the JSONL directory is already per-user (`~/.openhuman/users/<uuid>/workspace/...`), so cross-chat hits are naturally bounded to the requesting user. Active thread is excluded so we don't double-surface in-thread content.

Trade-offs:
- Substring-with-term-overlap scoring is intentionally simple and deterministic; the existing FTS5 path can graduate to primary once its sanitizer handles the punctuation cases that currently crash live queries.
- Per-turn full directory scan is acceptable at current thread counts; future work can add a per-user index if it shows up in profiles.

Manually smoked: built `OpenHuman.app` (debug) from the worktree, started two threads as the same user with overlapping topics, switched to a third thread, observed `[memory_loader] cross-chat JSONL scan returned N hits (exclude=Some(\"thread-…\"))` and the agent answered using context from the other threads without delegating to the tools agent.

## Submission Checklist

> If a section does not apply to this change, mark the item as `N/A` with a one-line reason. Do not delete items.

- [x] Tests added or updated (happy path + at least one failure / edge case) per [Testing Strategy](../gitbooks/developing/testing-strategy.md#failure-path-requirement)
- [x] **Diff coverage ≥ 80%** — changed lines (Vitest + cargo-llvm-cov merged via `diff-cover`) meet the gate enforced by [`.github/workflows/coverage.yml`](../.github/workflows/coverage.yml). Run `pnpm test:coverage` and `pnpm test:rust` locally; PRs below 80% on changed lines will not merge.
- [x] N/A: behaviour-only change to existing memory pipeline — Coverage matrix updated — added/removed/renamed feature rows in [`docs/TEST-COVERAGE-MATRIX.md`](../docs/TEST-COVERAGE-MATRIX.md) reflect this change (or `N/A: behaviour-only change`)
- [x] N/A: behaviour-only change — All affected feature IDs from the matrix are listed in the PR description under `## Related`
- [x] No new external network dependencies introduced (mock backend used per [Testing Strategy](../gitbooks/developing/testing-strategy.md#mock-policy))
- [x] N/A: does not touch release-cut surfaces — Manual smoke checklist updated if this touches release-cut surfaces ([`docs/RELEASE-MANUAL-SMOKE.md`](../docs/RELEASE-MANUAL-SMOKE.md))
- [x] Linked issue closed via `Closes #NNN` in the `## Related` section

## Impact

- Runtime: desktop. Per-turn JSONL scan adds bounded I/O (one open per thread file under the user's workspace). Term-overlap scoring is O(terms × messages); capped via `CROSS_CHAT_LIMIT * 4` shortlist then truncated to `CROSS_CHAT_LIMIT`.
- Security / privacy: scan is rooted at the active user's workspace directory; no cross-user enumeration. Active-thread exclusion prevents in-thread leakage into the cross-chat block.
- Migration: none — no schema changes, no new on-disk artifacts.
- Backward compat: callers that don't pass `workspace_dir` fall back to the prior FTS5 path; existing tests continue to exercise that branch via `MockMemory`.

## Related

- Closes #1505
- Follow-up PR(s)/TODOs: graduate FTS5 sanitizer (`sanitize_fts_query`) so the episodic path stops failing on live queries — see existing `episodic search failed: fts5: syntax error near "."` log noise; once green, the JSONL primary can flip back to FTS5 + the JSONL scan can serve as a fallback.

---

## AI Authored PR Metadata (required for Codex/Linear PRs)

> Keep this section for AI-authored PRs. For human-only PRs, mark each field `N/A`.

### Linear Issue
- Key: N/A
- URL: N/A

### Commit & Branch
- Branch: feat/1505-cross-chat-context
- Commit SHA: 2586cdc1

### Validation Run
- [x] `pnpm --filter openhuman-app format:check`
- [x] `pnpm typecheck`
- [x] Focused tests: `cargo test --lib openhuman::memory::conversations::store_tests::search_cross_thread_messages`
- [x] Rust fmt/check (if changed):
- [x] Tauri fmt/check (if changed):

### Validation Blocked
- `command:` N/A
- `error:` N/A
- `impact:` N/A

### Behavior Changes
- Intended behavior change: agent prompt now includes a `[Cross-chat context]` block populated from the same user's other JSONL chat threads, ranked by query-term overlap.
- User-visible effect: questions that depend on facts shared in a different chat by the same user are now answered without re-asking.

### Parity Contract
- Legacy behavior preserved: callers without `workspace_dir` continue to hit the FTS5 cross-session path via `Memory::recall(cross_session=true)`.
- Guard/fallback/dispatch parity checks: active-thread exclusion via `thread_context::current_thread_id()`; `min_relevance_score` honoured on both paths; `CROSS_CHAT_LIMIT` cap unchanged.

## Summary by CodeRabbit

* **New Features**
  * Added cross-chat context that surfaces relevant memories from previous conversation threads within the same workspace.
  * Implemented relevance-based filtering and snippet truncation for cross-chat memories to ensure concise, contextual information.
  * Deployed privacy-preserving provenance tracking that masks raw thread identifiers while maintaining source attribution.

[Review Change Stack](https://app.coderabbit.ai/change-stack/tinyhumansai/openhuman/pull/2054?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: src/openhuman/memory/store/memory_trait.rs [hunk 1]
```diff
@@ -171,7 +171,89 @@ impl Memory for UnifiedMemory {
...
+                let matched_count = query_terms
+                    .iter()
+                    .filter(|term| content_lower.contains(*term))
+                    .count();
+                if matched_count == 0 {
+                    // FTS5 surfaced a porter-stemmed match that doesn't
+                    // line up with our literal substring scan — keep the
+                    // row but assign a small floor score so it sorts
+                    // below substring matches without being dropped.
+                    let ts_rfc3339 = timestamp_to_rfc3339(entry.timestamp);
```
Comment (graycyrus) [inline src/openhuman/memory/store/memory_trait.rs]: **[major]** The floor score `0.1_f64.max(min_score)` defeats the downstream `min_relevance_score` gate. When `min_score` is the typical 0.4, this computes `floor = 0.4`, so porter-stemmed matches with zero literal query-term overlap always pass the `score >= min_relevance_score` filter in `memory_context.rs` and `memory_loader.rs`. That's the opposite of "sorts below substring matches" — it guarantees they survive.

Should be a fixed small value that naturally gets filtered by downstream score checks:

```rust
let floor = 0.05; // below typical min_relevance_score, filtered downstream
```

Or explicitly skip entries with zero literal matches rather than assigning them a score that's designed to pass.

File: src/openhuman/agent/memory_loader.rs [hunk 2]
```diff
@@ -15,6 +20,25 @@ const PRIOR_CONVERSATION_LIMIT: usize = 3;
...
+const CROSS_CHAT_LOADER_SNIPPET_CHARS: usize = 200;
+
+/// Render a short, non-leaky provenance tag for a cross-chat hit. The
+/// channel-side `session_id` is a JSON blob (`{"client_id": "...",
+/// "thread_id": "..."}`) — render only a short stable hash so the prompt
+/// never carries the raw `client_id` or socket UUID. See #1505 and the
+/// "Redact paths and IDs in public artifacts" auto-memory.
+fn provenance_tag(session_id: &str) -> String {
+    use std::hash::{Hash, Hasher};
+    let mut hasher = std::collections::hash_map::DefaultHasher::new();
```
Comment (graycyrus) [inline src/openhuman/agent/memory_loader.rs]: **[major]** `provenance_tag()` is duplicated verbatim from `memory_context.rs:22`. Same hash logic, same formatting — if the hashing strategy changes (e.g. for collision resistance or output format), both copies must be updated in lockstep.

Extract to a shared location — `crate::openhuman::util` or a `mod privacy` in the memory crate would be natural homes:

```rust
// e.g. in src/openhuman/util.rs or memory/privacy.rs
pub fn provenance_tag(session_id: &str) -> String { ... }
```

File: src/openhuman/agent/harness/memory_context.rs [hunk 3]
```diff
@@ -70,6 +103,51 @@ pub(crate) async fn build_context(
...
+    // the current chat is excluded by passing `session_id` so the block
+    // never duplicates the same-chat history.
+    let current_thread_id = crate::openhuman::providers::thread_context::current_thread_id();
+    let cross_session_opts = crate::openhuman::memory::RecallOpts {
+        session_id: current_thread_id.as_deref(),
+        cross_session: true,
+        min_score: Some(min_relevance_score),
+        ..Default::default()
+    };
+    if let Ok(entries) = mem
```
Comment (graycyrus) [inline src/openhuman/agent/harness/memory_context.rs:115]: **[major]** This entire cross-chat block has no debug logging. The sibling code in `memory_loader.rs` properly logs cross-chat scan results (`[memory_loader] cross-chat JSONL scan returned {} hits`) and budget exhaustion, and the surrounding codebase pattern (see `fork_context.rs`, `bus.rs`) uses `LOG_PREFIX` constants with `log::debug!` for entry/exit and branch decisions.

At minimum, log the recall result count and how many survived filtering:

```rust
log::debug!(
    "[memory-context] cross-chat recall returned {} entries, {} after filtering",
    entries.len(),
    cross.len(),
);
```

File: src/openhuman/agent/memory_loader.rs [hunk 4]
```diff
@@ -15,6 +20,25 @@ const PRIOR_CONVERSATION_LIMIT: usize = 3;
 /// do not auto-pollute every fresh chat.
 const PRIOR_CONVERSATION_KEY_PREFIX: &str = "high.";
 
+/// Per-snippet cap on the `[Cross-chat context]` block. Mirrors the
+/// harness-side helper but capped tighter for the loader path because
+/// loader runs against every turn (vs. build_context which is
+/// orchestrator-side).
+const CROSS_CHAT_LOADER_SNIPPET_CHARS: usize = 200;
+
+/// Render a short, non-leaky provenance tag for a cross-chat hit. The
```
Comment (graycyrus) [inline src/openhuman/agent/memory_loader.rs]: **[minor]** `CROSS_CHAT_LOADER_SNIPPET_CHARS = 200` here vs `CROSS_CHAT_SNIPPET_CHARS = 240` in `memory_context.rs`. The comment explains "capped tighter for the loader path", but the same cross-chat content will render at different lengths depending on which code path runs — confusing for debugging and inconsistent UX. Consider using one shared constant, or at least documenting why the difference matters in practice.

**Review by graycyrus — CHANGES_REQUESTED** [review]
## Walkthrough

Solid feature implementation for cross-chat context retrieval (#1505). The two-path design (JSONL primary + FTS5 fallback) is well-considered, privacy constraints are correctly enforced, and test coverage is decent for the fallback path. However, there are a few project-pattern violations and a scoring logic bug that need attention before merge. CodeRabbit's 4 actionable findings are valid and not repeated here — this review focuses on project-specific issues CodeRabbit missed.

## Change Summary

| File | Change type | Description |
|------|-------------|-------------|
| `memory_context.rs` | Modified | Adds `CROSS_CHAT_LIMIT`, `provenance_tag()`, `shorten_for_cross_chat()`, cross-chat block in `build_context()`, 5 new tests |
| `builder.rs` | Modified | Wires `.with_workspace_dir()` into `DefaultMemoryLoader` construction |
| `memory_loader.rs` | Modified | Adds `workspace_dir` field, duplicated `provenance_tag()`, JSONL primary + fallback cross-chat paths, budget-aware rendering, 4 new tests |
| `conversations/mod.rs` | Modified | Re-exports `CrossThreadHit` |
| `conversations/store.rs` | Modified | Adds `search_cross_thread_messages()` with substring scoring |
| `conversations/store_tests.rs` | Modified | 3 new cross-thread search tests |
| `conversations/types.rs` | Modified | New `CrossThreadHit` struct |
| `agentmemory/backend.rs` | Modified | Logs `cross_session` flag, explicit no-op comment |
| `memory_trait.rs` | Modified | Cross-session episodic recall in `UnifiedMemory::recall()`, 4 new tests |
| `fts5.rs` | Modified | `episodic_cross_session_search()`, `sanitize_fts_query()`, 5 new tests |
| `traits.rs` | Modified | `cross_session: bool` field on `RecallOpts`, documentation |

## Test Coverage Gap

**[major]** The primary JSONL scan path in `memory_loader.rs` (when `workspace_dir` is `Some`) has **zero** unit test coverage. All 4 loader tests exercise only the fallback `Memory::recall` path because they use `MockMemory` without setting `workspace_dir`. The primary path's budget-aware rendering, snippet truncation, and provenance tagging are production-critical but untested at the loader integration level. The `ConversationStore` tests cover the search itself, but not the loader's processing of its results.

Suggestion: Add at least one test that sets `workspace_dir` to a `tempdir`, writes JSONL thread files via `ConversationStore`, and verifies the loader's rendered output.

File: src/openhuman/agent/harness/memory_context.rs [hunk 5]
```diff
@@ -70,6 +103,51 @@ pub(crate) async fn build_context(
...
+        if !cross.is_empty() {
+            context.push_str("[Cross-chat context]\n");
+            for entry in &cross {
+                let prov = entry
+                    .session_id
+                    .as_deref()
+                    .map(provenance_tag)
+                    .unwrap_or_else(|| "chat:unknown".to_string());
+                let snippet = shorten_for_cross_chat(&entry.content);
+                let _ = writeln!(context, "- [{prov}] {snippet}");
```
[the remark this answers is not in the document]
Comment (oxoxDev) [reply src/openhuman/agent/harness/memory_context.rs:146]: Addressed in e2b433f9 — seen_keys.insert() in [User working memory] loop.

File: src/openhuman/agent/memory_loader.rs [hunk 6]
```diff
@@ -246,6 +284,112 @@ impl MemoryLoader for DefaultMemoryLoader {
...
+                })
+                .take(CROSS_CHAT_LIMIT)
+                .map(|e| {
+                    let sid = e
+                        .session_id
+                        .clone()
+                        .unwrap_or_else(|| "unknown".to_string());
+                    (sid, e.content)
+                })
+                .collect()
```
[the remark this answers is not in the document]
Comment (oxoxDev) [reply src/openhuman/agent/memory_loader.rs:362]: Addressed in 8b9bc872 — normalize session_id JSON blob before exclude compare.

File: src/openhuman/memory/conversations/store.rs [hunk 7]
```diff
@@ -132,6 +133,70 @@ impl ConversationStore {
...
+        let mut hits: Vec<CrossThreadHit> = Vec::new();
+        for thread in threads {
+            if exclude_thread_id == Some(thread.id.as_str()) {
+                continue;
+            }
+            let path = self.thread_messages_path(&thread.id);
+            let messages = match read_jsonl::<ConversationMessage>(&path) {
+                Ok(m) => m,
+                Err(_) => continue,
+            };
```
[the remark this answers is not in the document]
Comment (oxoxDev) [reply src/openhuman/memory/conversations/store.rs:179]: Addressed in b998cefe — tracing::warn! with path+error on Err arm.

File: src/openhuman/memory/store/unified/fts5.rs [hunk 8]
```diff
@@ -182,6 +182,116 @@ pub fn episodic_search(
...
+            .query_map(rusqlite::params![phrase_query, limit as i64], map_row)?
+            .collect::<Result<Vec<_>, _>>()?,
+    };
+
+    tracing::debug!(
+        "[fts5] cross-session search '{}' (exclude={:?}) returned {} results",
+        trimmed,
+        exclude_session,
+        rows.len()
+    );
```
[the remark this answers is not in the document]
Comment (oxoxDev) [reply src/openhuman/memory/store/unified/fts5.rs:279]: Addressed in 4280a9f8 — log query_hash + query_len, drop raw 'trimmed'.

File: src/openhuman/memory/store/memory_trait.rs (hunk 1 above)
Comment (oxoxDev) [reply src/openhuman/memory/store/memory_trait.rs]: Addressed in 244a4129 — skip zero-overlap FTS5 rows outright (drop floor logic).

File: src/openhuman/agent/memory_loader.rs (hunk 2 above)
Comment (oxoxDev) [reply src/openhuman/agent/memory_loader.rs]: Addressed in 916d02f2 — extracted to crate::openhuman::util::provenance_tag.

File: src/openhuman/agent/harness/memory_context.rs (hunk 3 above)
Comment (oxoxDev) [reply src/openhuman/agent/harness/memory_context.rs:115]: Addressed in 4ce8fa95 — tracing::debug! '[memory-context]' on entries vs filtered count.

File: src/openhuman/agent/memory_loader.rs (hunk 4 above)
Comment (oxoxDev) [reply src/openhuman/agent/memory_loader.rs]: Addressed in 916d02f2 — consolidated to shared CROSS_CHAT_SNIPPET_CHARS (240).

**Comment by oxoxDev** [comment]
Review back-fill (covers items that surfaced in review bodies rather than as inline threads):

- **CR nit** — `MockMemory::recall` in `memory_context.rs` tests now respects `RecallOpts.session_id` when `cross_session=true`, so the active-thread exclusion at `build_context` is actually exercised by the existing tests. Addressed in `4ce8fa95`.
- **graycyrus walkthrough — test coverage gap for primary JSONL path** — added `loader_surfaces_jsonl_primary_path_with_workspace_dir` in `memory_loader.rs` tests. Builds a tempdir-backed `ConversationStore`, wires `with_workspace_dir`, seeds one fact in a sibling thread, asserts the prompt renders `[Cross-chat context]` + matched snippet + `chat:<hash>` provenance + no raw `thread_id` leak. Addressed in `77cc9859`.

Also pushed `2ff581bd` for rustfmt auto-fixes from the pre-push hook.

File: src/openhuman/memory/conversations/store.rs [hunk 9]
```diff
@@ -132,6 +133,77 @@ impl ConversationStore {
         read_jsonl::<ConversationMessage>(&self.thread_messages_path(thread_id))
     }
 
+    /// Substring-match messages across **every** thread in the workspace,
+    /// optionally excluding one thread (the active chat). Returns up to
+    /// `limit` of the most-recent matching messages, newest first.
```
Comment (graycyrus) [inline src/openhuman/memory/conversations/store.rs:138]: **[major]** `search_cross_thread_messages` acquires `CONVERSATION_STORE_LOCK` at the top and holds it for the entire I/O-bound scan across every JSONL thread file. This method runs on every user message (via `load_context` → `DefaultMemoryLoader`), so in a workspace with many threads the global lock blocks all concurrent conversation writes (appends, thread creation) for the full scan duration.

The JSONL files are append-only, so reading them without the lock is safe — worst case you miss the very latest appended message, which is fine for cross-chat context. A deleted-mid-scan thread would hit the existing `Err(err) => continue` path.

Suggestion: acquire the lock only for `list_threads_unlocked()`, release it, then do the JSONL reads outside the lock:

```rust
let threads = {
    let _guard = CONVERSATION_STORE_LOCK.lock();
    self.list_threads_unlocked()?
};
// File reads happen without holding the lock
for thread in threads { ... }
```

File: src/openhuman/memory/store/unified/fts5.rs [hunk 10]
```diff
@@ -182,6 +182,126 @@ pub fn episodic_search(
...
+        trimmed.chars().count(),
+        exclude_session,
+        rows.len()
+    );
+    Ok(rows)
+}
+
+/// Best-effort FTS5 query sanitiser: strip characters that break the
+/// MATCH grammar (quotes, parens, asterisks) and wrap each surviving
+/// whitespace-delimited token in double quotes so the FTS engine treats
```
Comment (graycyrus) [inline src/openhuman/memory/store/unified/fts5.rs:285]: **[minor]** Magic number `.take(8)` — extract to a named constant (e.g. `const MAX_FTS_QUERY_TOKENS: usize = 8;`) for self-documentation. Other modules in this file use named constants for similar caps.

File: src/openhuman/util.rs [hunk 11]
```diff
@@ -2,6 +2,25 @@
...
 //! This module contains reusable helper functions used across the codebase.
 
+/// Render a short, non-leaky provenance tag for a session/thread id.
+///
+/// The channel-side `session_id` is typically a JSON blob
+/// (`{"client_id": "...", "thread_id": "..."}`); rendering it verbatim
+/// in a model prompt or log line would leak the raw `client_id` /
+/// socket UUID. Hash the input with `DefaultHasher` and emit only the
+/// low 32 bits as `chat:xxxxxxxx` — short, stable per id, and not
+/// reversible to the original blob.
```
Comment (graycyrus) [inline src/openhuman/util.rs:12]: **[minor]** `DefaultHasher` output is explicitly not stable across Rust toolchain versions (per stdlib docs). Currently this is only used for prompt rendering (not persisted or matched across sessions), so it's safe — but a brief comment noting the non-stability would prevent future misuse as a persistent key.

Suggestion: add a one-liner like `// Note: DefaultHasher is not stable across Rust versions — safe here since tags are prompt-only, not persisted.`

**Review by graycyrus — DISMISSED** [review]
## Walkthrough

Continuation review after 10 fix/refactor/test commits addressing all 5 prior findings + all 5 CodeRabbit findings. All previously flagged issues are resolved — nice work. The feature is well-structured with 21 new tests, proper provenance hashing, and clean backward compatibility.

One new concern: the JSONL cross-thread scan holds the global conversation store lock for its entire I/O-bound pass, which could block concurrent writes in active workspaces. Two minor style nits below.

## Change Summary

| File | Change type | Description |
|------|-------------|-------------|
| `memory_context.rs` | Modified | Cross-chat block in `build_context()` with provenance tags, snippet caps, score filtering; 4 new tests |
| `session/builder.rs` | Modified | Wires `config.workspace_dir` into `DefaultMemoryLoader` |
| `memory_loader.rs` | Modified | JSONL primary path + Memory-trait fallback cross-chat retrieval; 5 new tests including integration-style JSONL round-trip |
| `conversations/mod.rs` | Modified | Re-export `CrossThreadHit` |
| `conversations/store.rs` | Modified | `search_cross_thread_messages()` — substring scan across threads |
| `conversations/store_tests.rs` | Modified | 3 new cross-thread search tests (hit finding, exclusion, edge cases) |
| `conversations/types.rs` | Modified | `CrossThreadHit` struct definition |
| `agentmemory/backend.rs` | Modified | No-op `cross_session` handling with logging |
| `memory_trait.rs` | Modified | Cross-session episodic recall in `UnifiedMemory`; 4 new tests |
| `unified/fts5.rs` | Modified | `episodic_cross_session_search()`, `sanitize_fts_query()`; 5 new tests |
| `traits.rs` | Modified | `RecallOpts.cross_session` field + docs |
| `util.rs` | Modified | `provenance_tag()` function |

## Prior findings status (Review 1)

All 5 findings addressed:
- ✅ Floor score defeating downstream gate → zero-overlap hits now dropped (`if matched_count == 0 { continue; }`)
- ✅ `provenance_tag()` duplication → extracted to `crate::openhuman::util::provenance_tag`
- ✅ Zero debug logging in cross-chat block → `tracing::debug!("[memory-context] ...")` added
- ✅ Primary JSONL path untested → `loader_surfaces_jsonl_primary_path_with_workspace_dir` test added
- ✅ Inconsistent snippet caps → unified to `CROSS_CHAT_SNIPPET_CHARS=240`

**Review by graycyrus — COMMENTED** [review]
## Review 3 — continuation (same commit as Review 2)

No new commits since my last review. Checking in on the three outstanding findings.

### Prior REQUEST_CHANGES — still open

All three findings from Review 2 remain unaddressed:

1. **[major] `store.rs:138`** — `search_cross_thread_messages` holds `CONVERSATION_STORE_LOCK` for the entire I/O-bound JSONL scan. This blocks all concurrent conversation writes while scanning every thread file. Consider releasing the lock after `list_threads_unlocked()` returns the thread list, then scanning files outside the lock (the JSONL files are append-only so a lockless read is safe for substring search).

2. **[minor] `fts5.rs:285`** — Magic number `.take(8)` in `sanitize_fts_query` should be a named constant (e.g., `MAX_FTS_QUERY_TOKENS`).

3. **[minor] `util.rs:12`** — `DefaultHasher` is not stable across Rust versions. Since `provenance_tag` output appears in prompts only (not persisted), this is acceptable — but add a one-line comment noting the stability contract so future maintainers don't start persisting these hashes.

### New observation

4. **[minor]** Logging pattern deviation in `store.rs`: the new `search_cross_thread_messages` uses `tracing::warn!` with `[conversations]` prefix, but the rest of `store.rs` uses `log::warn!` with the `LOG_PREFIX` constant (`[memory:conversations]`). Sibling code (`bus.rs`) also uses `log::` exclusively. Align to `log::warn!("{LOG_PREFIX} ...")` for grep consistency within this module.

### Resolved threads

All 5 findings from Review 1 were addressed in prior fix commits. Resolved the remaining open thread (debug logging in `memory_context.rs`). The prior REQUEST_CHANGES from Review 2 remains blocking until the lock contention issue is addressed.

---

@oxoxDev — the lock contention in `search_cross_thread_messages` (finding #1) is the main blocker. The two minor items can ride along. Let me know if you have questions about the lock-free approach.

**Review by graycyrus — APPROVED** [review]
